## Vehicle Application EDA

**Select Menu By Car Model**

We will analyze the car data to organize the information for a select menu where customers can choose the car model they are interested in and view the information about each of those models available that are for sale. This will be accompanied by a slider that isolates the car models by the year they were created.

**Price Analysis**

We will also analyze the car data by creating a histogram and a scatterplot to show the distribution of price differences based on important factors such as car age, odometer (mileage), the vehicles condition, type and fuel requirements.

## Data Pre-Processing

In [4]:
import pandas as pd
import numpy as np

In [6]:
df = pd.read_csv('../vehicles_us.csv')

In [8]:
df.head() #get a general overview of the data

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28

In [10]:
df.info() #get a general overview of the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


In [12]:
df['model'] = df['model'].str.replace(r'[\s-]', '_', regex=True) #removes whitespace and hypens from 'model' column.

In [14]:
df['type'] = df['type'].str.lower() #lowercase the 'type' column

In [16]:
df = df.dropna().reset_index(drop=True) #drop missing values and reset the index while dropping the old index

In [18]:
df['date_posted'] = pd.to_datetime(df['date_posted']) #convert 'date_posted' column to datetimeIndex type

In [20]:
df[['model_year','odometer','cylinders']] = df[['model_year','odometer','cylinders']].astype(int)  #convert float columns to integers

In [22]:
df.duplicated().sum() #no duplicate rows

0

In [24]:
df.head() #check that changes have been applied

price  model_year                     model  condition  cylinders fuel  \
0  14990        2014              chrysler_300  excellent          6  gas   
1  15990        2013               honda_pilot  excellent          6  gas   
2  19500        2011  chevrolet_silverado_1500  excellent          8  gas   
3  12990        2009                 gmc_yukon  excellent          8  gas   
4  14990        2010                  ram_1500  excellent          8  gas   

   odometer transmission    type paint_color  is_4wd date_posted  days_listed  
0     57954    automatic   sedan       black     1.0  2018-06-20           15  
1    109473    automatic     suv       black     1.0  2019-01-07           68  
2    128413    automatic  pickup       black     1.0  2018-09-17           38  
3    132285    automatic     suv       black     1.0  2019-01-31           24  
4    130725    automatic  pickup         red     1.0  2018-12-30           13

In [26]:
df.info() #check dtypes for date and integer changes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14852 entries, 0 to 14851
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   price         14852 non-null  int64         
 1   model_year    14852 non-null  int64         
 2   model         14852 non-null  object        
 3   condition     14852 non-null  object        
 4   cylinders     14852 non-null  int64         
 5   fuel          14852 non-null  object        
 6   odometer      14852 non-null  int64         
 7   transmission  14852 non-null  object        
 8   type          14852 non-null  object        
 9   paint_color   14852 non-null  object        
 10  is_4wd        14852 non-null  float64       
 11  date_posted   14852 non-null  datetime64[ns]
 12  days_listed   14852 non-null  int64         
dtypes: datetime64[ns](1), float64(1), int64(5), object(6)
memory usage: 1.5+ MB


We discovered that there were 13 columns with 51525 rows. After removing all of the missing values 14852 rows remained. After that we cleaned up the data to remove whitespace and hypens. Then, made sure all of the data was lowercased to make sure we can spot duplicates easily.  

## Data Analysis

In [30]:
car_model = sorted(df['model'].unique()) #extract the types of car models

In [32]:
car_model #view the unique value of car models

['acura_tl',
 'bmw_x5',
 'buick_enclave',
 'cadillac_escalade',
 'chevrolet_camaro',
 'chevrolet_colorado',
 'chevrolet_corvette',
 'chevrolet_cruze',
 'chevrolet_equinox',
 'chevrolet_malibu',
 'chevrolet_silverado',
 'chevrolet_silverado_1500',
 'chevrolet_silverado_1500_crew',
 'chevrolet_silverado_2500hd',
 'chevrolet_silverado_3500hd',
 'chevrolet_suburban',
 'chevrolet_tahoe',
 'chevrolet_trailblazer',
 'chevrolet_traverse',
 'chrysler_200',
 'chrysler_300',
 'chrysler_town_&_country',
 'dodge_charger',
 'dodge_dakota',
 'dodge_grand_caravan',
 'ford_edge',
 'ford_escape',
 'ford_expedition',
 'ford_explorer',
 'ford_f150',
 'ford_f150_supercrew_cab_xlt',
 'ford_f250',
 'ford_f250_super_duty',
 'ford_f350',
 'ford_f350_super_duty',
 'ford_f_150',
 'ford_f_250',
 'ford_f_250_sd',
 'ford_f_250_super_duty',
 'ford_f_350_sd',
 'ford_focus',
 'ford_fusion',
 'ford_fusion_se',
 'ford_mustang',
 'ford_ranger',
 'ford_taurus',
 'gmc_acadia',
 'gmc_sierra',
 'gmc_sierra_1500',
 'gmc_sierr

In [34]:
min_year, max_year = int(df['model_year'].min()), int(df['model_year'].max()) #get min and max model_year

In [36]:
min_year, max_year

(1965, 2019)

We needed to analysis the data to extract the unique vehicle models and the years those models were created in order to create an application select box and slider for our customers to navigate by model_year and car_model. These navigation features will allow them to access all of the data based off their selected criteria.

## Feature Engineering

In [40]:
def age_category(x):      #create a function to define by age
    if x<5: return '<5'
    elif x>=5 and x<10: return '5-10'
    elif x>=10 and x<20: return '10-20'
    else: return '>20'

In [42]:
df['age'] = 2024 - df['model_year'] #calculate the age of the cars

In [44]:
df['age_category'] = df['age'].apply(age_category) #apply the function to the data

In [46]:
df.head() #view the data to make sure the function applied

price  model_year                     model  condition  cylinders fuel  \
0  14990        2014              chrysler_300  excellent          6  gas   
1  15990        2013               honda_pilot  excellent          6  gas   
2  19500        2011  chevrolet_silverado_1500  excellent          8  gas   
3  12990        2009                 gmc_yukon  excellent          8  gas   
4  14990        2010                  ram_1500  excellent          8  gas   

   odometer transmission    type paint_color  is_4wd date_posted  days_listed  \
0     57954    automatic   sedan       black     1.0  2018-06-20           15   
1    109473    automatic     suv       black     1.0  2019-01-07           68   
2    128413    automatic  pickup       black     1.0  2018-09-17           38   
3    132285    automatic     suv       black     1.0  2019-01-31           24   
4    130725    automatic  pickup         red     1.0  2018-12-30           13   

   age age_category  
0   10        10-20  
1   11        10-20  
2   13        10-20  
3   15        10-20  
4   14        10-20

We wanted to create an application for customers to navigate price distributions especially by the age of the car. So we created a function for customers to help them analysis the differences in these features easier.